### PR TITLE
Don't use JobQueue during shutdown (RIPD-1356):

### PR DESCRIPTION
**Pull request by scottschurr**
If the `JobQueue` is used during shutdown then those `Jobs` may access `Stoppables` after they have already stopped.  This violates the preconditions of `Stoppables` and may lead to undefined behavior.

The solution taken here is to reference count all `Jobs` in the `JobQueue`.  At stop time all `Jobs` already in the `JobQueue` are allowed to run to completion, but no further `Jobs` are allowed into the `JobQueue`.

If a `Job` is rejected from the `JobQueue` (because we are stopping), then `JobQueue::addJob()` returns false, so the caller can make any necessary adjustments.

The pull request is divided into three different commits:

1. The first commit does two things.  First, when the `JobCounter` was originally introduced I thought each `Stoppable` would have a `JobCounter`.  That ended up being impractical.  So I changed the model to having a single `JobCounter` which is held by the `RootStoppable`.  This `JobCounter` is used to guarantee that all counted `Job`s are completed before any of the `Stoppables` begin stopping.  The rest of the commit switches most calls of `JobQueue::addJob()` into `JobQueue::addCountedJob()`.  The concern here is that all calls to `addJob()` previously were assumed to always be successful.  Now the call may fail (returning `false`) once the system begins shutting down.  It seems that in _most_ cases it's okay if adding the `Job` fails while we're shutting down.  In those specific cases where we need the operation to complete then I added code to do the right thing.  As code reviewers you get to ponder whether I added the right code in all of the necessary places.

2. The second commit deals with `JobQueue::Coro`.  It turns out that `JobQueue::Coro` objects are executed by adding them as `Job`s to the `JobQueue`.  Handling a `Coro` when a `Job` might not be successfully added to the `JobQueue` ended up being a bit tricky.  So that conversion is in its own commit.  This commit converts all of the remaining `JobQueue::addJob()` calls to `JobQueue::addCountedJob()`.

3. The last commit tidies up the naming.  Since `JobQueue::addJob()` is no longer a used name, we can convert our new longer name `addCountedJob()` back to `addJob()`.  It's no longer useful to distinguish between counted and uncounted `Job`s -- all `Job`s are counted.

My biggest concern during the code review is whether I missed any places where we need recovery code if `addJob()` fails (and returns `false`).

Assignees: @miguelportilla, @bachase.  More are welcome.  The more eyes the better.

### CODE REVIEW

File: src/ripple/rpc/handlers/RipplePathFind.cpp [hunk 1]
```diff
@@ -55,9 +55,95 @@ Json::Value doRipplePathFind (RPC::Context& context)
...
+        //    Fortunately this problem can be addressed by looking at the
+        //    return value of makeLegacyPathRequest().  If
+        //    makeLegacyPathRequest() cannot get a thread to run the path-find
+        //    on, then it returns an empty request.
+        //
+        // 2. The path-finding job might run, but the Coro::post() might be
+        //    rejected by the JobQueue (because we're shutting down).
+        //
+        //    We handle this case by:
+        //      a. setting the successfulPost flag in the Coro to false, then
```
Comment (bachase) [inline src/ripple/rpc/handlers/RipplePathFind.cpp]: Where is `successfulPost` flag defined?

File: src/ripple/core/impl/Stoppable.cpp [hunk 2]
```diff
@@ -113,12 +107,13 @@ void Stoppable::stopAsyncRecursive (beast::Journal j)
     auto const start = high_resolution_clock::now();
     onStop ();
     auto const ms = duration_cast<milliseconds>(
```
Comment (bachase) [inline src/ripple/core/impl/Stoppable.cpp]: :man_facepalming: I missed that.  No need for change.

File: src/ripple/core/Coro.ipp [hunk 3]
```diff
@@ -48,7 +48,9 @@ inline
 JobQueue::Coro::
 ~Coro()
 {
+#ifndef NDEBUG
```
Comment (bachase) [inline src/ripple/core/Coro.ipp]: No need if that helps clarify `finished_`.

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 4]
```diff
@@ -1200,48 +1201,51 @@ LedgerMaster::updatePaths (Job& job)
...
 */
-void
+bool
 LedgerMaster::newPFWork (const char *name)
 {
     if (mPathFindThread < 2)
     {
-        ++mPathFindThread;
-        app_.getJobQueue().addJob (
+        if (app_.getJobQueue().addJob (
```
Comment (bachase) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp]: 👍 Would be nice to have that type check at compile-time, but clearly outside the scope of this effort.

File: src/ripple/app/main/NodeStoreScheduler.cpp [hunk 5]
```diff
@@ -48,16 +46,21 @@ void NodeStoreScheduler::onChildrenStopped ()
...
+        // Recover by executing the task synchronously.
+        doTask (task);
+    }
 }
 
 void NodeStoreScheduler::doTask (NodeStore::Task& task)
 {
     task.performScheduledTask ();
-    if ((--m_taskCount == 0) && isStopping())
+    if ((--m_taskCount == 0) && isStopping() && areChildrenStopped())
```
Comment (bachase) [inline src/ripple/app/main/NodeStoreScheduler.cpp]: That diagram is extremely useful.  I'd consider adding it to the code somewhere, even if it might get stale.

I'll need to study this section a bit more as well and am not advocating a change yet.

Comment (bachase) [inline src/ripple/app/main/NodeStoreScheduler.cpp]: Reading the [Stoppable documentation](https://github.com/ripple/rippled/blob/develop/src/ripple/core/Stoppable.h#L125-L163), I suspect the `NodeStoreScheduler` was written to just follow the example code there.  I don't think `NodeStoreScheduler` has any extra asynchronous logic outside of the JobQueue tasks, which is why `NodeStoreScheduler::onChildrenStopped` has `assert(m_taskCount == 0)`.  If you imagined `m_taskCount` were counting tasks in some local thread pool (rather than the child `Stoppable` `JobQueue`), then `NodeStoreScheduler::onChildrenStopped` could only call `stopped` if the imagined internal thread pool were empty at that time.  If it were not empty and `doTask` was servicing that thread pool, it would be responsible for calling `stopped()` once the last task finished.   Therefore, I think your fix is a good one, as you should not be calling `stopped` if you don't know your children have stopped.

So TL;DR, I think your fix is a good one to match the general `Stoppable` framework. However, I suspect given that `NodeStoreScheduler` has internal logic to wait on, we could probably rewrite `doTask` as below and leave `NodeStoreScheduler::onChildrenStopped` responsible for signaling `stopped`.
```
void NodeStoreScheduler::doTask (NodeStore::Task& task)
{
    task.performScheduledTask ();
}
```
I'm still not advocating heavily for such a change, just want to put my thoughts down.

Comment (scottschurr) [inline src/ripple/app/main/NodeStoreScheduler.cpp]: I agree with your assessment on both counts:

1. The diagram is useful and should be somewhere in the source code even if it gets stale.  I've toyed with a couple of places to leave the diagram.  Application.cpp is where a lot of this structure gets built.  So somewhere in there almost makes sense.  But I'm leaning toward putting the diagram in Stoppable.h.  This is a little weird.  All of the elements of the diagram are `Stoppable`, but `Stoppable` itself is ignorant of the teaming classes below that it serves.  Nevertheless, I think putting the diagram in Stoppable.h improves the discoverability of the diagram.  What are your thoughts?

2. I agree with your assessment that, if there are no bugs, it should never be possible for `doTask` to call `stopped()`.  I'm hesitant to make such a change in this particular pull request.  However I'd like to add a comment in the `doTask()` source including your assessment.  This might encourage someone in the future to study the situation and potentially remove the call to `stopped()`.

**Comment by scottschurr** [comment]
Pushed a commit that I believe addresses @bachase's very useful comments.

File: src/ripple/app/ledger/Ledger.cpp [hunk 6]
```diff
@@ -1048,25 +1048,22 @@ bool pendSaveValidated (
...
-    else
-    {
-        app.getJobQueue ().addJob(
-            jtPUBOLDLEDGER, "Ledger::pendOldSave", job);
+        return true;
     }
 
-    return true;
+    // The JobQueue won't do the Job.  Do the save synchronously.
+    return saveValidatedLedger(app, ledger, isCurrent);
```
Comment (miguelportilla) [inline src/ripple/app/ledger/Ledger.cpp]: `JobType` and `jobName` are unnecessarily created if `isSynchronous`.

Alternate suggestion:
```
    // Do the save synchronously.
    if (isSynchronous)
        return saveValidatedLedger(app, ledger, isCurrent);

    JobType const jobType {isCurrent ? jtPUBLEDGER : jtPUBOLDLEDGER};
    char const* const jobName {
        isCurrent ? "Ledger::pendSave" : "Ledger::pendOldSave"};
    return app.getJobQueue().addJob (jobType, jobName,
        [&app, ledger, isCurrent] (Job&) {
            saveValidatedLedger(app, ledger, isCurrent);
        });
```

File: src/ripple/app/misc/Validations.cpp [hunk 7]
```diff
@@ -429,25 +429,34 @@ class ValidationsImp : public Validations
...
+                mWriting = true;
+                doWrite (sl);
+            }
 
-        while (mWriting)
-        {
-            ScopedUnlockType sul (mLock);
-            std::this_thread::sleep_for (std::chrono::milliseconds (100));
+            // Handle the case where flush() is called while a queuedWrite
+            // is already in progress.
```
Comment (miguelportilla) [inline src/ripple/app/misc/Validations.cpp]: I am not sure I understand the comment. It seems that if queudWrite was in progress it would hold the lock and we could never arrive here at the same time.

File: src/ripple/app/misc/Validations.cpp [hunk 8]
```diff
@@ -456,22 +465,27 @@ class ValidationsImp : public Validations
         if (mWriting)
             return;
 
-        mWriting = true;
-        app_.getJobQueue ().addJob (
-            jtWRITE, "Validations::doWrite",
-            [this] (Job&) { doWrite(); });
+        // addCountedJob() may return false (Job not added) at shutdown.
+        mWriting = app_.getJobQueue ().addCountedJob (
```
Comment (miguelportilla) [inline src/ripple/app/misc/Validations.cpp]: Assigning `mWriting` the return value of addJob may be inaccurate. We may prefer to have `doWrite` be the only place that assigns `mWriting`. True upon entry and false on exit. That simplifies the code a bit and is precise on when we are writing.

File: src/ripple/app/misc/Validations.cpp [hunk 9]
```diff
@@ -429,25 +429,34 @@ class ValidationsImp : public Validations
...
+                    anyNew = true;
+                }
+            }
+            mCurrentValidations.clear ();
 
-        if (anyNew)
-            condWrite ();
+            // If there isn't a write in progress already, then write to the
+            // database synchronously.
+            if (anyNew && !mWriting)
```
Comment (miguelportilla) [inline src/ripple/app/misc/Validations.cpp]: A race condition exists here and in `condWrite`. Say `mWriting` is true but doWrite is done and assigning `mWiting` false. We expect a flush but it won't occur until Flush gets called again and if we are shutting down, Flush may have been called for the last time.

File: src/ripple/app/misc/Validations.cpp [hunk 10]
```diff
@@ -429,25 +429,34 @@ class ValidationsImp : public Validations
...
+            {
+                mWriting = true;
+                doWrite (sl);
+            }
```
Comment (miguelportilla) [inline src/ripple/app/misc/Validations.cpp]: Shouldn't this return at this point? doWrite just assigned `mWriting` false.

File: src/ripple/app/misc/Validations.cpp [hunk 11]
```diff
@@ -456,22 +465,27 @@ class ValidationsImp : public Validations
...
+        // addCountedJob() may return false (Job not added) at shutdown.
+        mWriting = app_.getJobQueue ().addCountedJob (
+            jtWRITE, "Validations::queuedWrite",
+            [this] (Job&) { queuedWrite(); });
     }
 
-    void doWrite ()
+    void queuedWrite ()
     {
         auto event = app_.getJobQueue ().makeLoadEvent (jtDISK, "ValidationWrite");
```
Comment (miguelportilla) [inline src/ripple/app/misc/Validations.cpp]: `event` is unused.

File: src/ripple/app/misc/Validations.cpp (hunk 8 above)
Comment (miguelportilla) [inline src/ripple/app/misc/Validations.cpp]: Thank you for the detailed explanation, that was helpful. I rechecked your work and it looks good.

`mWriting` seems to be serving two purposes. One purpose determines if a write job has been dispatched. The other determines if writing to the database is currently in progress. I suppose for clarity it could be broken up into two bools but I am not suggesting you make that change, the current code works.

File: src/ripple/app/misc/Validations.cpp (hunk 9 above)
Comment (scottschurr) [inline src/ripple/app/misc/Validations.cpp]: I will go ahead and pass that lock to `condWrite()`.  This stuff is hard enough to think about that we shouldn't make it unnecessarily obscure.

File: src/ripple/app/misc/Validations.cpp (hunk 11 above)
Comment (scottschurr) [inline src/ripple/app/misc/Validations.cpp]: `event` is an odd duck.  We're instantiating a `LoadEvent` here, which is a scoped timing kind of thing.  It adds information to the LoadMonitor when it is created and destroyed.  So it's doing work, it just doesn't look like it 😃.

**Comment by scottschurr** [comment]
Rebased to 0.80.0-b1.  The only complication to the rebase is that Validations.cpp was deleted, so I needed to verify that my changes were properly ported to RCLValidations.cpp.  I think I've made the only necessary small change, but I'd appreciate if @bachase would look over that change here:  https://github.com/ripple/rippled/pull/2135/files#diff-9647085a1d5d0d2cf7b0b835b0e8c742L62

If @bachase is good with that change then I'll squash and mark the pull request as passed.

**Comment by bachase** [comment]
👍  on the rebased changes in `RCLValidations`.

**Comment by scottschurr** [comment]
I'm closing this passed pull request, but only because it is now incorporated into https://github.com/ripple/rippled/pull/2145, which depends on these changes.  This will hopefully make it easier for whoever assembles the next beta version.
